api-rights-bridge/src/test/java/org/xwiki/contrib/rights/internal/DefaultRightsWriterTest.java

```diff
@@ -409,6 +410,42 @@ void testSaveRuleWithNullRights() throws XWikiException
         assertEquals(1, rightsObj.getIntValue(ALLOW_PROPERTY));
     }
 
+    @Test
+    void testSaveRuleWithEmptyRightsList() throws XWikiException
+    {
+        DocumentReference documentReference = new DocumentReference("xwiki", "S", "P");
+        // prepare rules to put on the document, with empty rights set
+        WritableSecurityRule rule = new WritableSecurityRuleImpl(
+            Collections.singletonList(new DocumentReference("xwiki", "XWiki", "XWikiAdminGroup")),
+            Collections.emptyList(), new RightSet(Right.VIEW), RuleState.ALLOW);
+
+        // call the function under test
+        this.rightsWriter.saveRules(Arrays.asList(rule), documentReference);
+
+        // get the document that was just modified and assert on it
+        XWikiDocument resultDoc =
+            this.oldcore.getSpyXWiki().getDocument(documentReference, this.oldcore.getXWikiContext());
+        // check that there is a single object set
+        assertEquals(1, resultDoc.getXObjects(DefaultRightsWriter.XWIKI_RIGHTS_CLASS).size());
+        BaseObject rightsObj = resultDoc.getXObject(DefaultRightsWriter.XWIKI_RIGHTS_CLASS);
+        // check that the allow is set to default true
+        assertEquals(1, rightsObj.getIntValue(ALLOW_PROPERTY));
+        // and the rest of the object is set properly, as we asked for
+        assertEquals("", rightsObj.getLargeStringValue(USERS_PROPERTY));
+        assertEquals("XWiki.XWikiAdminGroup", rightsObj.getLargeStringValue(GROUPS_PROPERTY));
+        assertEquals(1, rightsObj.getIntValue(ALLOW_PROPERTY));
+
+        // Clean all rights objects
+        this.rightsWriter.saveRules(Collections.emptyList(), documentReference);
+
+        resultDoc =
+            this.oldcore.getSpyXWiki().getDocument(documentReference, this.oldcore.getXWikiContext());
+
+        // check that the right object was removed
+        // Note the object still be in the map, but it's set to null.
+        assertNull(resultDoc.getXObject(DefaultRightsWriter.XWIKI_RIGHTS_CLASS));
```

Comment:
Actually, this is more or less an implementation detail of setting rules on a terminal page (document) not on a non-terminal page (space) and specifically the fact that when we do this, the saveRules() function doesn't actually save the storage document for the rules, _and_ the objects array in XWiki will leave null holes.

What we expect is that there is no null object, actually, when rules are saved on a non-terminal page (space). Maybe we should have 2 different tests, one for terminal page and one for non-terminal page.

Reply:
I don't really understand the difference between terminal and non-terminal pages about object in array which still have null.

From what I understand when we remove an object on a document, it's set to null on the object list of the document. So if we remove an object on a terminal page or on a `WebPreferences` page (in case of non-terminal page), as I understand in both cases we will have a null in the list.

Can you clarify this or correct me if I pass over something ?

Note I checked for null because I didn't find a other better way to check that the right was correctly removed.
If you have a better way to check this, I would be glade to use it.

Reply:
> From what I understand when we remove an object on a document, it's set to null on the object list of the document. So if we remove an object on a terminal page or on a WebPreferences page (in case of non-terminal page), as I understand in both cases we will have a null in the list.

Yes, when an object is removed from a page, there is a null put in its place, and the trailing nulls are cleaned up on save of that document. 

Otherwise put, if you have an object of class A, you delete it, you call getObjects() on the document java object you have, you get an array with one position, and the value on that position is null. If you save the document, and then you call getIObjects() again, you get an array with 0 positions and no null in it.
If you have a document with two objects of class A, objects 0 and 1 and you delete the first one, getObjects() on the java object you have for the document will return an array with 2 positions, first one is null second one is non-null. If you save the document and then call getObjects() again, you get the same array with 2 positions of which the first one is null and the second one is an object. This is how objects arrays work, for reasons that are long to explain here (you should find some comments in the code, though, I think).

Now, since we're using storage mocks for tests instead of actual databases, maybe storage mocks are not implementing this objects management exactly like this, but this is how it happens in the database.

So what is really different here between terminal and non-terminal is that:
* rights on non-terminal pages are not stored in the page itself but in the WebPreferences terminal page child of that page.
* rights on terminal pages are stored on the page itself.

Reply:
Not 100% sur about what you was expecting on this. I tried to push a improvement on this side.

Reply:
> So what is really different here between terminal and non-terminal is that:
    rights on non-terminal pages are not stored in the page itself but in the WebPreferences terminal page child of that page.
    rights on terminal pages are stored on the page itself.

That's not entirely true: it depends the scope of the right. e.g. if I go to the Main.WebHome page (a non terminal page) and from the admin I edit the right of that page only (Users & Rights > Rights: Page), the right object is added to the page itself. It's only if I edit the right of the space that (Users & Rights > Rights: Page & Children) that the right is saved to the WebPreference page. 
Now I haven't checked exactly the scenario here, but just to mention that it's not enough for the page to be non terminal for the right to be saved in WebPreference.